added feature to copy track/operation into other new track/operation …

joernu76 opened the pull request:
…without using files

**Purpose of PR?**:

Fixes #2699

**Does this PR introduce a breaking change?**
no

**If the changes in this PR are manually verified, list down the scenarios covered:**:
manually checked all combinations

**Checklist:**
- [ ] Bug fix. Fixes #<issue number>
- [x] New feature (Non-API breaking changes that adds functionality)
- [x] PR Title follows the convention of  `<type>: <subject>`
- [ ] Commit has unit tests

<!--

The PR title message must follow convention:
`<type>: <subject>`.

Where: <br />
- `type` is to define what type of PR is this.
  Most common types are:
    - `feat`      - for new features, not a new feature for build script
    - `fix`       - for bug fixes or improvements, not a fix for build script
    - `chore`     - changes not related to production code
    - `docs`      - changes related to documentation
    - `style`     - formatting, missing semi colons, linting fix etc; no significant production code changes
    - `test`      - adding missing tests, refactoring tests; no production code change
    - `refactor`  - refactoring production code, eg. renaming a variable or function name, there should not be any significant production code changes

- `subject` is a single line brief description of the changes made in the pull request.

-->

joernu76 committed:
added feature to copy track/operation into other new track/operation without using files

joernu76 commented:
I see no "easy" way to import waypoints from a non-active operation? Or am I mistaken?

joernu76 commented:
@ReimarBauer

ReimarBauer commented:
See the multiple flightpath dockingwidget.
You can fetch without activation.

ReimarBauer commented:
See hhe multiple flightpath dockingwidget.
You can fetch without activation.

ReimarBauer commented:
And the autoplot cli can fetch too.

ReimarBauer commented:
When I do here Copy from selected, what should happen?
 
<img width="579" alt="Bildschirmfoto 2025-03-20 um 21 31 37" src="https://github.com/user-attachments/assets/9065886e-9547-473e-8af5-b88706eba749" />

I get "ERROR: cannot copy into oneself"

In that case maybe the entry should be disabled

ReimarBauer commented:
It is meant, one is activated and the other is selected to "send" the data in this case to new revision
<img width="736" alt="Bildschirmfoto 2025-03-20 um 21 39 23" src="https://github.com/user-attachments/assets/cc6b2700-3a5e-42d2-a3ef-b71c60306bb3" />

In the revision that is not "named" but we should do that. 

This enables to destinguish regular edits.

ReimarBauer commented on mslib/msui/mscolab.py:
```diff
@@ -1030,8 +1047,9 @@ def add_operation(self):
-            data["content"] = self.add_proj_dialog.f_content
+        if f_content is not None:
+            data["content"] = f_content
+        print(data)
```
this could become a logging or removed?

joernu76 replied:
should be removed.

ReimarBauer commented:
maybe longer "Copy from selected to activated"

ReimarBauer commented:
when an flightpath get sended to an mscolab OP we maybe should create a new named revision. 
When multiple people send, those can be listed filtered because of the name.

We should maybe also define a different service message for the chat.

It would be more important to have a different service message in the chat. Because that is the log what happens in the operation. Names can be removed the log is persistent.

ReimarBauer commented on mslib/msui/mscolab.py:
```diff
@@ -985,29 +990,41 @@ def browse():
         self.proj_diag.show()
 
+    def add_operation_from_new_dialogue(self):
+        logging.debug("add_operation_from_dialogue")
```
the name is difficult later to understand or to map to the origin. maybe use same writing (add_operation_from_new_dialog)

self.add_proj_dialog.buttonBox.accepted.connect(self.add_operation_from_new_dialogue)

ReimarBauer commented on mslib/msui/mscolab.py:
```diff
@@ -952,8 +954,11 @@ def delete_own_account(self, _=None):
+        self.add_operation_dialogue()
+
+    @verify_user_token
+    def add_operation_dialogue(self, name=None, description=None, xml=None):
```
we wrote always dialog

that can become later a wording refactoring task

joernu76 replied:
done.

ReimarBauer reviewed (changes requested):
a few things to change

and I don't see a problem to have it in a 10.1.0 when you send it to stable. In that case I would skip 10.0.2 and go directly to 10.1.0

ReimarBauer commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -735,6 +738,85 @@ def handle_export_local(self, extension, function, pickertype):
+            name=name + f" (copy {self.new_flight_track_counter:d})")
+        # Make a copy of the template. Otherwise, all new flight tracks would
+        # use the same data structure in memory.
+        waypoints_model.insertRows(0, rows=len(template_copy), waypoints=template_copy)
```
here we could add a flag which writes then in the chat a message, e.g. "copied data". The chat system messages can be seen as a log what happens to the operation 

https://github.com/Open-MSS/MSS/blob/develop/mslib/msui/flighttrack.py#L439

joernu76 commented:
> maybe longer "Copy from selected to activated"

I am not happy with the terminology either. for practical reasons, one end-point need be the "active" flight track, so we must copy from selected to activated. I adopt your proposal. One could also make this an "import" feature...?

EDIT: I moved it into "Import From". Tell me what you think about it.

joernu76 commented:
> a few things to change
> 
> and I don't see a problem to have it in a 10.1.0 when you send it to stable. In that case I would skip 10.0.2 and go directly to 10.1.0

I do not remember our reasoning here. Shouldn't new features go to develop and be made into a release from there?
Ie. develop -> x+1.0.0 or x.y+1.0 (depending on API breakage) and stable -> x.y.z+1?

joernu76 commented:
> When I do here Copy from selected, what should happen?
> <img alt="Bildschirmfoto 2025-03-20 um 21 31 37" width="579" src="https://private-user-images.githubusercontent.com/278728/425173741-9065886e-9547-473e-8af5-b88706eba749.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI4MDQzNDUsIm5iZiI6MTc0MjgwNDA0NSwicGF0aCI6Ii8yNzg3MjgvNDI1MTczNzQxLTkwNjU4ODZlLTk1NDctNDczZS04YWY1LWI4ODcwNmViYTc0OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyNFQwODE0MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jM2Y0MDYxNDUzNTEwZWZlZWY2NmFiZjFiZGNjNDM3N2JjZjc3NzM0MDI0NGRiNDEyMjZmYjExZTU5Y2I4YjYwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.cSUWc4B9m267Op0uBOCUjtVPJePcbiJ8UrA363IYYBg">
> 
> I get "ERROR: cannot copy into oneself"
> 
> In that case maybe the entry should be disabled

I decided an the "easy" error message to get it done instead of an easy-to-get-wrong activating/deactivating. I will look into how difficult this really is. We currently have no triggers at all on stuff being simply "selected". An here it is not only being selected, but what is selected...

joernu76 commented:
> It is meant, one is activated and the other is selected to "send" the data in this case to new revision <img alt="Bildschirmfoto 2025-03-20 um 21 39 23" width="736" src="https://private-user-images.githubusercontent.com/278728/425176832-cc6b2700-3a5e-42d2-a3ef-b71c60306bb3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI4MDQzNDUsIm5iZiI6MTc0MjgwNDA0NSwicGF0aCI6Ii8yNzg3MjgvNDI1MTc2ODMyLWNjNmIyNzAwLTNhNWUtNDJkMi1hM2VmLWI3MWM2MDMwNmJiMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyNFQwODE0MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wYTc1YzExY2Y5MGQ2YzA1YmUyMmRjZTY5OTc4YmQ5MzU4YjRlY2FhOGQ3NDcxNTRlZWFlOTdlNzU3MjFkNGJkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Ihh8drc7rLoJ7nM_adpteZWrtJs8pXANC8GKXq3s7Ug">
> 
> In the revision that is not "named" but we should do that.
> 
> This enables to destinguish regular edits.

I do not follow, what you request here? The current behaviour is similar to an "import flightrack" for operations and thus consistent. Adding a revision/chat message might be a good idea, but should be consistently applied.

joernu76 commented:
I fixed the simple things. (dialogue, moved copy-from-selected to "import" menu.
1) wrt to chat-message/named revision, I propose to move this into an "easy for beginners issue".
2) wrt to disabling the entry when the active-flighttrack/operation is selected, I propose to keep it as it is. Doing this difficult to get right and adds a lot of unnecessary code. I think people will manage. Otherwise -> new issue

I thought develop is the proper branch for a non-bug-fix change, but this can be merged into stable as well.

joernu76 committed:
changes according to discussion

ReimarBauer commented:
> > a few things to change
> > and I don't see a problem to have it in a 10.1.0 when you send it to stable. In that case I would skip 10.0.2 and go directly to 10.1.0
> 
> I do not remember our reasoning here. Shouldn't new features go to develop and be made into a release from there? Ie. develop -> x+1.0.0 or x.y+1.0 (depending on API breakage) and stable -> x.y.z+1?

when features don't break existing code we can have a minor release. When the API changes it must be a major.

joernu76 committed:
update savechangedmessage

joernu76 committed:
fixed pytest errors

ReimarBauer reviewed (changes requested):
besides that chat message everything fine.

I get only:
![Bildschirmfoto 2025-03-24 um 21 06 07](https://github.com/user-attachments/assets/9806ddb6-f945-4b68-b9b1-da3d4b08f6dc)

please have a look on that

when that needs more changes we can do it later.

joernu76 commented:
Same here. So merge?

ReimarBauer commented:
> Same here. So merge?

yes, we can. I am just waiting until at least the ubuntu test succeeds.

joernu76 committed:
fix more tests

joernu76 commented:
Wit the Mac-Tests failing I didn't notice the failing Ubuntu test. Might work now.

ReimarBauer reviewed (approved):
Thx
